.ansible-lint-ignore

```
@@ -8,3 +8,9 @@ ansible/roles/stig/tasks/TOSS_04_020160.yml fqcn[action-core]
 # below rules are ignored in order to grep the appropriate configuration asked in TOSS STIG item 030900 since using a ansible task shell module does not allow a pipe command
 ansible/roles/stig/tasks/TOSS_04_030900.yml command-instead-of-shell
 ansible/roles/stig/tasks/TOSS_04_030900.yml risky-shell-pipe
+
+# below rules ignore "failed_when" so when tasks fail, the ansible playbook will continue
```

I'm not sure I understand why these rules are ignoring errors.

---

([copying from mattermost post](https://mattermost.energy.gov/lc-llnl/pl/5kha5fb75tge8bagzd3f138iyc))

I spoke with @cdefrates and we agree that the lesser evil would be to have the `ignore_errors: true` for these specific needs, at least until we find a better method. One I had are to change them to `ansible.builtin.debug` so it prints out the failed statement, but the task result shows up as "ok". Another idea I had was to add a final task to each `TOSS-04-*.yml` file that says whether said stig passed or not, but I'm worried about runtime considering that means adding 227 more tasks AND having to register a variable to all subtasks that we care about to justify whether a toss stig item passed or not.

For details on our tests on `failed_when: ${some_conditional}`, `failed_when: false`, `ignore_errors: true`, and `quiet: true`, see below:


- for items that have some ansible task with `failed_when: ${some conditional}` and the ansible task failed, the playbook stops. For example, the task is:
```
      ansible.builtin.command: "grep hosts /etc/nsswitch.conf"
      changed_when: false
      check_mode: false
      failed_when: dns_entry.rc not in [0,1]
      register: dns_entry
```
.
		- when `/etc/nsswitch.conf` does not exist, the playbook ends:
```
TASK [toss_stig : TOSS-04-010330 - Determine whether the system is using local or DNS name resolution] ************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "cmd": ["grep", "hosts", "/etc/nsswitch.conf"], "delta": "0:00:00.008253", "end": "2024-12-10 13:51:06.446974", "failed_when_result": true, "msg": "non-zero return code", "rc": 2, "start": "2024-12-10 13:51:06.438721", "stderr": "grep: /etc/nsswitch.conf: No such file or directory", "stderr_lines": ["grep: /etc/nsswitch.conf: No such file or directory"], "stdout": "", "stdout_lines": []}

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0
```
.
	- for the same thing, but the `failed_when` conditional is met, it stops the playbook from completing:
```
TASK [toss_stig : TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty] ***************************************************************************************************************
fatal: [localhost]: FAILED! => {"assertion": "resolv_stat.stat.size == 0", "changed": false, "evaluated_to": false, "msg": "If local host authentication is being used, the /etc/resolv.conf file must be emtpy"}

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0
```
.
- if I add a `failed_when: false` to a task that has a `failed_when: ${some conditional}`, a warning appears AND the task shows up as okay even though it should have failed:
```
TASK [Check for incompatible operating systems] *******************************************************************************************************************************************
ok: [localhost]
**[WARNING]: While constructing a mapping from /etc/ansible/roles/toss_stig/tasks/TOSS_04_010330.yml, line 20, column 7, found a duplicate dict key (failed_when). Using last defined value**
**only.**
included: /etc/ansible/roles/toss_stig/tasks/TOSS_04_010330.yml for localhost

TASK [toss_stig : TOSS-04-010330 - Determine whether the system is using local or DNS name resolution] ************************************************************************************
ok: [localhost]
```
.
- if I remove the conditional and only have `failed_when: false`, the task shows up as "ok". The below example is when `/etc/nsswitch.conf` does not exist so it should show up with an error message that the file does not exist:
```
TASK [toss_stig : TOSS-04-010330 - Determine whether the system is using local or DNS name resolution] ************************************************************************************
ok: [localhost]
```
.
- for items that have use `ansible.builtin.assert` to only fail based on a conditional and has `quiet:true`, the playbook stops running. For example, say I have this task:
```
    - name: TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty
      ansible.builtin.assert:
        that:
          - resolv_stat.stat.size == 0
        fail_msg: "If local host authentication is being used, the /etc/resolv.conf file must be emtpy"
        quiet: true
```
.
		- the playbook shows the fatal message and stops running:
```
TASK [toss_stig : TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty] ***************************************************************************************************************
fatal: [localhost]: FAILED! => {"assertion": "resolv_stat.stat.size == 0", "changed": false, "evaluated_to": false, "msg": "If local host authentication is being used, the /etc/resolv.conf file must be emtpy"}

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0
```
.
- say I add `failed_when: false` to the task, the task results in "ok" nor does it print the fail message wanted, and continues with the playbook:
```
TASK [toss_stig : TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty] ***************************************************************************************************************
ok: [localhost]

TASK [toss_stig : TOSS-04-010330 - Verify the operating system is configured to use two or more name servers for DNS resolution] **********************************************************
skipping: [localhost]
```
.
	- if `ignore_errors: true` is added, the task results show "failed" and the error message and continues the playbook, but `...ignoring` follows the message and the recap counts it as a "skipped" task:
```
TASK [toss_stig : TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty] ***************************************************************************************************************
fatal: [localhost]: FAILED! => {"assertion": "resolv_stat.stat.size == 0", "changed": false, "evaluated_to": false, "msg": "If local host authentication is being used, the /etc/resolv.conf file must be emtpy"}
...ignoring

TASK [toss_stig : TOSS-04-010330 - Verify the operating system is configured to use two or more name servers for DNS resolution] **********************************************************
skipping: [localhost]

TASK [toss_stig : TOSS-04-010330 - Check that there are at least two name servers for DNS resolution] *************************************************************************************
fatal: [localhost]: FAILED! => {"assertion": "num_name_servers.stdout is defined", "changed": false, "evaluated_to": false, "msg": "Configure the operating system to use two or more name servers for DNS resolution"}
...ignoring

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0    skipped=2    rescued=0    **ignored=2**
```

---

ok